File: serialize_and_deserialize_binary_tree.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Codec:
    def serialize(self, root):
        if not root:
            return ""

        result = []
        queue = deque([root])

        while queue:
            node = queue.popleft()

            if node:
                result.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append("null")

        return ",".join(result)

    def deserialize(self, data):
        if not data:
            return None

        values = data.split(",")
        root = TreeNode(int(values[0]))
        queue = deque([root])
        index = 1

        while queue and index < len(values):
            node = queue.popleft()

            if values[index] != "null":
                node.left = TreeNode(int(values[index]))
                queue.append(node.left)
            index += 1

            if index < len(values) and values[index] != "null":
                node.right = TreeNode(int(values[index]))
                queue.append(node.right)
            index += 1

        return root


if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)

    codec = Codec()

    serialized_data = codec.serialize(root)
    print("Serialized:", serialized_data)

    deserialized_root = codec.deserialize(serialized_data)
    print("Root after deserialization:", deserialized_root.val)

File: Day-25/serialize_and_deserialize_binary_tree.cpp
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int value) {
        val = value;
        left = NULL;
        right = NULL;
    }
};

class Codec {
public:
    string serialize(TreeNode* root) {
        if (root == NULL) {
            return "";
        }

        queue<TreeNode*> q;
        q.push(root);
        string result = "";

        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();

            if (node == NULL) {
                result += "null,";
            } else {
                result += to_string(node->val) + ",";
                q.push(node->left);
                q.push(node->right);
            }
        }

        return result;
    }

    TreeNode* deserialize(string data) {
        if (data.empty()) {
            return NULL;
        }

        stringstream ss(data);
        string value;
        getline(ss, value, ',');

        TreeNode* root = new TreeNode(stoi(value));
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();

            if (!getline(ss, value, ',')) break;
            if (value != "null") {
                node->left = new TreeNode(stoi(value));
                q.push(node->left);
            }

            if (!getline(ss, value, ',')) break;
            if (value != "null") {
                node->right = new TreeNode(stoi(value));
                q.push(node->right);
            }
        }

        return root;
    }
};

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(4);
    root->right->right = new TreeNode(5);

    Codec codec;

    string serializedTree = codec.serialize(root);
    cout << "Serialized Tree: " << serializedTree << endl;

    TreeNode* deserializedRoot = codec.deserialize(serializedTree);
    cout << "Root after Deserialization: " << deserializedRoot->val << endl;

    return 0;
}
